src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/RollupMetadataService.kt

```diff
@@ -15,12 +15,407 @@
+                .allowPartialSearchResults(false)
+            val response: SearchResponse = client.suspendUntil { search(searchRequest, it) }
+
+            if (response.hits.hits.isEmpty()) {
+                // Empty doc hits will result in a no-op from the runner
+                return StartingTimeResult.NoDocumentsFound
+            }
+
+            val firstSource = response.hits.hits.first().sourceAsMap[dateHistogram.sourceField] as String
+
+            return StartingTimeResult.Success(getRoundedTime(firstSource, dateHistogram))
+        } catch (e: RemoteTransportException) {
+            val unwrappedException = ExceptionsHelper.unwrapCause(e) as Exception
+            logger.debug("Error when getting initial start time for rollup [${rollup.id}]: $unwrappedException")
+            return StartingTimeResult.Failure(unwrappedException)
+        } catch (e: Exception) {
+            // TODO: Catching general exceptions for now, can make more granular
+            logger.debug("Error when getting initial start time for rollup [${rollup.id}]: $e")
+            return StartingTimeResult.Failure(e)
+        }
+    }
+
+    /**
+     * Return time rounded down to the nearest unit of time the interval is based on.
+     * This should map to the equivalent bucket a document with the given timestamp would fall into for the date histogram.
+     */
+    private fun getRoundedTime(timestamp: String, dateHistogram: DateHistogram): Instant {
+        val roundingStrategy = getRoundingStrategy(dateHistogram)
+        val timeInMillis = ZonedDateTime.parse(timestamp).toInstant().toEpochMilli()
+
+        val roundedMillis = roundingStrategy
+            .prepare(timeInMillis, timeInMillis)
+            .round(timeInMillis)
+        return Instant.ofEpochMilli(roundedMillis)
+    }
+
+    /** Takes an existing start or end time and returns the value for the next window based on the rollup interval */
+    private fun getShiftedTime(time: Instant, rollup: Rollup): Instant {
+        val dateHistogram = rollup.dimensions.first() as DateHistogram
+        val roundingStrategy = getRoundingStrategy(dateHistogram)
+
+        val timeInMillis = time.toEpochMilli()
+        val nextRoundedMillis = roundingStrategy
+            .prepare(timeInMillis, timeInMillis)
+            .nextRoundingValue(timeInMillis)
+        return Instant.ofEpochMilli(nextRoundedMillis)
+    }
+
+    /**
+     * Get the rounding strategy for the given time interval in the DateHistogram.
+     * This is used to calculate time windows by rounding the given time based on the interval.
+     */
+    // TODO: Could make this an extension function of DateHistogram and add to some utility file
+    private fun getRoundingStrategy(dateHistogram: DateHistogram): Rounding {
+        val intervalString = (dateHistogram.calendarInterval ?: dateHistogram.fixedInterval) as String
+        // TODO: Make sure the interval string is validated before getting here so we don't get errors
```

Comment:
Let's double check this to confirm when is the earliest point in the code that we catch an incorrect DateHistogram interval string. If it isn't caught before here, we would enter the `else` statement and may not handle the case correctly.

Perhaps we could leverage the DateHistogram builder and try to enter the given interval string and catch any exceptions that are thrown. If this approach is valid, we could even put it in the `init` of `Rollup` to validate early.

Reply:
Will cover TODOs and any cleanups needed in follow up PRs to unblock the PRs that depend on these.